Snakefile

```diff
@@ -178,6 +395,152 @@ rule trimmomatic_merge:
+  """
+  input:
+    "fastqc_after_trim"
+  output:
+    directory("multiqc_after_trim")
+  log:
+    "logs/multiqc_after_trim.log"
+  params:
+    memory="4"
+  threads:
+    1
+  shell:
+    """
+    mkdir {output} &> {log}
+    multiqc -o {output} {input} &>> {log}
+    """
+
+rule compare_qc_after_trim:
+  """
+  Compares the quality of reads before and after trimming.
+  Compares PER BASE SEQUENCE QUALITY, PER SEQUENCE QUALITY SCORES, 
+  OVERREPRESENTED SEQUENCES and ADAPTER CONTENT modules of FastQC.
+  """
+  input:
+    before="multiqc_before_trim/multiqc_data/multiqc_fastqc.txt",
+    after="multiqc_after_trim/multiqc_data/multiqc_fastqc.txt"
+  output:
+    result="FastQC_comparison_after_trim.txt"
+  log:
+    "logs/compare_qc_after_trim.log"
+  run:
+ 
+    with open(input["before"], "r") as before, open(input["after"], "r") as after, open(output["result"], "w") as out_file:
+      before_data = [line.strip() for line in before.readlines()[1:]] # skip header
+      after_data = [line.strip() for line in after.readlines()[1:]] # skip header
+
+      per_base_sequence_quality_list = []
+      per_sequence_quality_scores_list = []
+      overrepresented_sequences_list = []
+      adapter_content_list = []
+
+      for line in before_data:
+        basic_statistics, per_base_sequence_quality, per_tile_sequence_quality, per_sequence_quality_scores, per_base_sequence_content, per_sequence_gc_content, per_base_n_content, sequence_length_distribution, sequence_duplication_levels, overrepresented_sequences, adapter_content = line.split("\t")[10:]
+        per_base_sequence_quality_list.append(per_base_sequence_quality)
+        per_sequence_quality_scores_list.append(per_sequence_quality_scores)
+        overrepresented_sequences_list.append(overrepresented_sequences)
+        adapter_content_list.append(adapter_content)
+    
+      per_base_sequence_quality_list_after = []
+      per_sequence_quality_scores_list_after = []
+      overrepresented_sequences_list_after = []
+      adapter_content_list_after = []
+  
+      for line in after_data:
+        basic_statistics_after, per_base_sequence_quality_after, per_tile_sequence_quality_after, per_sequence_quality_scores_after, per_base_sequence_content_after, per_sequence_gc_content_after, per_base_n_content_after, sequence_length_distribution_after, sequence_duplication_levels_after, overrepresented_sequences_after, adapter_content_after = line.split("\t")[10:]
+        per_base_sequence_quality_list_after.append(per_base_sequence_quality_after)
+        per_sequence_quality_scores_list_after.append(per_sequence_quality_scores_after)
+        overrepresented_sequences_list_after.append(overrepresented_sequences_after)
+        adapter_content_list_after.append(adapter_content_after)
+        
+      out_file.write("FASTQC COMPARISON BEFORE AND AFTER TRIMMING\n")
+      out_file.write('-' * 50 + '\n')
+
+      out_file.write("**PER BASE SEQUENCE QUALITY**\n")
```

Comment:
This code should not be duplicated (before trimmomatic and after trimmomatic). Can you find a way how to move the code to some function, so it can be called twice?

Reply:
True, I tried to make the code nicer, you can check the update :)